# Combining DataFrames With Pandas - Lab

## Introduction

In this lab, you'll gain practice combining DataFrames through concatenation.  You'll also practice executing various types of joins to selectively combine the information stored in the tables!

## Objectives

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)


## Getting Started

You'll start with a quick section to gain practice with concatenating datasets using `pd.concat()`.

## Concatenating DataFrames

Run the cell below to create some sample DataFrames for us to work with.  

In [2]:
import pandas as pd
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                    index=[0, 1, 2, 3])


df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                    index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'], 
                    'D': ['D8', 'D9', 'D10', 'D11']},
                    index=[8, 9, 10, 11])

Now that you have multiple DataFrames to work with, you can execute a concatenation to join them together.  

In the cell below, concatenate the 3 DataFrames together using the appropriate function.  

In [3]:
to_concat = [df1, df2, df3]
combined_df = pd.concat(to_concat)
combined_df.head(6)

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3
4  A4  B4  C4  D4
5  A5  B5  C5  D5

**_EXPECTED OUTPUT:_**

<img src="images/er1.png">

## Setting Join Conditions With Concatenation

You can also specify if the concatenation is an **_Outer Join_** or an **_Inner Join_**.  Next, you'll execute an inner join. Before you do, you need to create another table that contains some overlapping index values with a DataFrame that already exists.  

Run the cell below to create the new DataFrame.

In [4]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'],
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                    index=[2, 3, 6, 7])

Now, in the cell below, use the `pd.concat()` method to join DataFrames 1 and 4.  However, this time, specify that the `join` is `'inner'`, and `axis=1`. 

In [5]:
to_concat = [df1, df4]
df1_and_4 = pd.concat(to_concat, join = 'inner', axis = 1)
df1_and_4

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

**_EXPECTED OUTPUT:_**

<img src='images/er2.png'>

You'll notice that in this case, the results returned contain only the rows with indexes that exist in both tables--rows 2 and 3.  The resulting table contains the values for each column in both tables for the rows.  

Note that there are many, many ways that you can make full use of the `concat()` functionality in pandas to join DataFrames together--these are just a few of the most common examples pulled from the pandas documentation on the subject.  For a full view of all the ways you can use `pd.concat()`, see the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html)!

## Loading In Data
Now, it's time to move on to working with the Hearthstone cards database.  This database contains information on cards from the popular game, [Hearthstone](https://playhearthstone.com/en-us/)! For full information on the dataset, see the  [kaggle page](https://www.kaggle.com/jeradrose/hearthstone-cards) for this dataset. 

This database consists of the following tables:

* _cards_
* *dust_costs*
* _entourages_
* _mechanics_
* *play_requirements*

Many of rows in each table--but not all--correspond to the same cards. As such, each table contains a column called `card_id` which acts as a **_Primary Key_** for each table.  You'll make use of these keys to **_join_** the different tables together into a single DataFrame.  You'll also experiment with different types of joins to help us decide exactly what information you wish to combine.  

Simply run the cell below to import the tables from the database as DataFrames.

In [77]:
cards_df = pd.read_csv('cards.csv')
dust_df = pd.read_csv('dust.csv')
entourages_df = pd.read_csv('entourages.csv')
mechanics_df = pd.read_csv('mechanics.csv')
play_requirements_df = pd.read_csv('play_requirements.csv')

Great.  Now, let's set the correct column, `card_id`, as the index column for each of these tables, and then display each to ensure that everything is as expected.  

For each of the dataframes you created in the cell above, call the `.set_index()` method and pass in `card_id`.  Also set `inplace=True`.  Then, display the head of each respective DataFrame to ensure everything worked.  

**_NOTE:_** Since you are performing this operation in place, running any cell a second time will result in pandas throwing an error.  If you need to run something a second time, restart the kernel using the jupyter notebook menu at the top of the page.  

In [74]:
cards_df.index, dust_df.index, entourages_df.index, mechanics_df.index, play_requirements_df.index

(RangeIndex(start=0, stop=2819, step=1),
 RangeIndex(start=0, stop=4040, step=1),
 RangeIndex(start=0, stop=328, step=1),
 RangeIndex(start=0, stop=1079, step=1),
 RangeIndex(start=0, stop=776, step=1))

In [78]:
cards_df.set_index('card_id', inplace=True)
cards_df.head()

player_class    type                name      set  \
card_id                                                        
KARA_00_07         MAGE   SPELL       Astral Portal     KARA   
NEW1_008a         DRUID   SPELL   Ancient Teachings  EXPERT1   
BRM_010t2         DRUID  MINION  Druid of the Flame      BRM   
AT_132          NEUTRAL  MINION  Justicar Trueheart      TGT   
OG_141          NEUTRAL  MINION   Faceless Behemoth       OG   

                                                         text  cost  attack  \
card_id                                                                       
KARA_00_07           Summon a random <b>Legendary</b> minion.   1.0     NaN   
NEW1_008a                                        Draw a card.   0.0     NaN   
BRM_010t2                                                 NaN   3.0     2.0   
AT_132      <b>Battlecry:</b> Replace your starting Hero P...   6.0     6.0   
OG_141                                                    NaN  10.0    10.0   

            health     rarity  collectible  \
card_id                                      
KARA_00_07     NaN        NaN          NaN   
NEW1_008a      NaN        NaN          NaN   
BRM_010t2      5.0     COMMON          NaN   
AT_132         3.0  LEGENDARY          1.0   
OG_141        10.0     COMMON          1.0   

                                                       flavor   race  \
card_id                                                                
KARA_00_07                                                NaN    NaN   
NEW1_008a                                                 NaN    NaN   
BRM_010t2                                                 NaN  BEAST   
AT_132      It's like putting racing stripes and a giant s...    NaN   
OG_141      Rejected names: Forty-Foot Faceless, Big ol' N...    NaN   

           how_to_earn how_to_earn_golden targeting_arrow_text faction  \
card_id                                                                  
KARA_00_07         NaN                NaN                  NaN     NaN   
NEW1_008a          NaN                NaN                  NaN     NaN   
BRM_010t2          NaN                NaN                  NaN     NaN   
AT_132             NaN                NaN                  NaN     NaN   
OG_141             NaN                NaN                  NaN     NaN   

            durability  
card_id                 
KARA_00_07         NaN  
NEW1_008a          NaN  
BRM_010t2          NaN  
AT_132             NaN  
OG_141             NaN

In [79]:
dust_df.set_index('card_id', inplace=True)

In [80]:
entourages_df.set_index('card_id', inplace=True)

In [81]:
mechanics_df.set_index('card_id', inplace=True)

In [82]:
play_requirements_df.set_index('card_id', inplace=True)

## Executing Joins

Now that you have the tables loaded correctly, we're going to execute some joins. There are 4 different kinds of joins, which can best be visualized with venn diagrams:

<img src='images/joins.png'>

In these diagrams, each circle represents a DataFrame or SQL Table.  The left table is the table you are working with, and the right table is the table you want to join to the table you are working with.  You'll start by executing the most common type of join, an **_Inner Join_**.

In the cell below, join `cards_df` with `mechanics_df` using the built-in `.join()` method on the `cards_df` object. 

Pass in the following parameters:
* the table you want to join with, `mechanics_df`
* The `how` parameter set to the type of join you want, `'inner'`

In [83]:
cards_with_mechanics_df = cards_df.join(mechanics_df, how='inner', rsuffix='card_id')
cards_with_mechanics_df

player_class         type                        name  \
card_id                                                                      
AT_002                       MAGE        SPELL                      Effigy   
AT_005t                   NEUTRAL       MINION                        Boar   
AT_006                       MAGE       MINION            Dalaran Aspirant   
AT_007                       MAGE       MINION                Spellslinger   
AT_009                       MAGE       MINION                      Rhonin   
AT_010                     HUNTER       MINION                Ram Wrangler   
AT_011e                   NEUTRAL  ENCHANTMENT            Light's Blessing   
AT_012                     PRIEST       MINION            Spawn of Shadows   
AT_017                    NEUTRAL       MINION           Twilight Guardian   
AT_018                     PRIEST       MINION         Confessor Paletress   
AT_019                    WARLOCK       MINION                  Dreadsteed   
AT_022                    WARLOCK        SPELL            Fist of Jaraxxus   
AT_023                    WARLOCK       MINION                Void Crusher   
AT_028                      ROGUE       MINION             Shado-Pan Rider   
AT_030                      ROGUE       MINION           Undercity Valiant   
AT_032                      ROGUE       MINION                Shady Dealer   
AT_035t                     ROGUE        SPELL                     Ambush!   
AT_036                      ROGUE       MINION                   Anub'arak   
AT_037                      DRUID        SPELL                Living Roots   
AT_038                      DRUID       MINION          Darnassus Aspirant   
AT_038                      DRUID       MINION          Darnassus Aspirant   
AT_039                      DRUID       MINION            Savage Combatant   
AT_039e                     DRUID  ENCHANTMENT                      Savage   
AT_040                      DRUID       MINION                  Wildwalker   
AT_042                      DRUID       MINION          Druid of the Saber   
AT_042t                     DRUID       MINION             Sabertooth Lion   
AT_042t2                    DRUID       MINION          Sabertooth Panther   
AT_045                      DRUID       MINION                      Aviana   
AT_046                     SHAMAN       MINION             Tuskarr Totemic   
AT_047                     SHAMAN       MINION         Draenei Totemcarver   
...                           ...          ...                         ...   
TB_SPT_DPromoSecret3      WARRIOR        SPELL            Visions of Valor   
TB_SPT_DPromoSecret4      WARRIOR        SPELL             Visions of Fate   
TB_SPT_DPromoSecret5      WARRIOR        SPELL       Visions of the Amazon   
TB_SPT_DPromoSecret6      WARRIOR        SPELL     Visions of the Sorcerer   
TB_SPT_DPromoSecret7      WARRIOR        SPELL  Visions of the Necromancer   
TB_SPT_DPromoSecret9      WARRIOR        SPELL        Visions of Knowledge   
TB_SPT_DPromo_Hero2       WARRIOR         HERO                The Cow King   
TB_SPT_DpromoPortal       NEUTRAL       MINION            Enigmatic Portal   
TB_SPT_DpromoPortal       NEUTRAL       MINION            Enigmatic Portal   
TB_SPT_Minion1            NEUTRAL       MINION                  Shieldsman   
TB_SPT_Minion2            NEUTRAL       MINION             Battle Standard   
TB_SPT_Minion2            NEUTRAL       MINION             Battle Standard   
TU4c_003                  NEUTRAL       MINION                      Barrel   
TU4c_008e                 NEUTRAL  ENCHANTMENT              Might of Mukla   
TU4f_007                  NEUTRAL       MINION                Crazy Monkey   
XXX_001                   NEUTRAL        SPELL                    Damage 1   
XXX_002                   NEUTRAL        SPELL                    Damage 5   
XXX_008                   NEUTRAL        SPELL                      Freeze   
XXX_020                   NEUTRAL        SP

Examine the output from the cell above and compare it to the original output of both the `cards_df` and `mechanics_df` DataFrame heads you displayed earlier.  Notice how it this now combines the columns from both?

**_Question_**

If you inspect the original `cards_df` DataFrame, you'll notice that it contains  2,819 records.  The result of our inner join, `cards_with_mechanics_df`, conntains only 1079 rows.  Why?

Write your answer below this line: "Mechanis only has 1079 records"
________________________________________________________________________________

First performed an inner join, which only includes records that are present in both tables.  Although there were 2819 records in the left table, there were only 1079 records that existed in both tables, which are what you see in the resulting dataframe. 

## Other Types of Joins

By default, the `.join()` method performs a left join if no parameter is passed in for `how=`.  In the cell below, perform a **_Left Join_** of `cards_with_mechanics_df` and `play_requirements_df`, with `cards_with_mechanics_df` as the left table.  

Then, display `left_join_df` to inspect our results. 

In [51]:
left_join_df = cards_with_mechanics_df.join(play_requirements_df, rsuffix = 'card_id')
left_join_df

card_id player_class         type  \
0                     KARA_00_07         MAGE        SPELL   
1                      NEW1_008a        DRUID        SPELL   
2                      BRM_010t2        DRUID       MINION   
3                         AT_132      NEUTRAL       MINION   
4                         OG_141      NEUTRAL       MINION   
5                        CRED_17      NEUTRAL       MINION   
6                        OG_337e      NEUTRAL  ENCHANTMENT   
7                      BRMA08_1H      NEUTRAL         HERO   
8                       GVG_011a      NEUTRAL  ENCHANTMENT   
9                    NAX8_02H_TB      NEUTRAL   HERO_POWER   
10                        AT_131      NEUTRAL       MINION   
11                      CS2_122e      NEUTRAL  ENCHANTMENT   
12                   LOEA05_02ha      NEUTRAL   HERO_POWER   
13             CS2_102_H1_AT_132      WARRIOR   HERO_POWER   
14                      CFM_334e      NEUTRAL  ENCHANTMENT   
15                       EX1_583      NEUTRAL       MINION   
16        TB_PickYourFate_11rand      NEUTRAL        SPELL   
17                      LOE_007t      WARLOCK        SPELL   
18                       CFM_095      NEUTRAL       MINION   
19                       BRM_018      PALADIN       MINION   
20                      PRO_001a      NEUTRAL        SPELL   
21                     BRMA10_5H      NEUTRAL       MINION   
22                      BRMA11_1      NEUTRAL         HERO   
23                        OG_027       SHAMAN        SPELL   
24                       NAX1_01      NEUTRAL         HERO   
25                       GVG_030        DRUID       MINION   
26                       CFM_094      WARLOCK        SPELL   
27                   DS1h_292_H1       HUNTER   HERO_POWER   
28                       CS2_227      NEUTRAL       MINION   
29                       KAR_062      NEUTRAL       MINION   
...                          ...          ...          ...   
1049                      AT_017      NEUTRAL       MINION   
1050                    LOEA12_1      NEUTRAL         HERO   
1051                     GVG_035        DRUID       MINION   
1052                     OG_104e      NEUTRAL  ENCHANTMENT   
1053                     CS2_119      NEUTRAL       MINION   
1054                     LOE_003         MAGE       MINION   
1055                    LOE_113e      NEUTRAL  ENCHANTMENT   
1056                      OG_284      NEUTRAL       MINION   
1057                     EX1_611       HUNTER        SPELL   
1058                      OG_340      NEUTRAL       MINION   
1059                      OG_323      NEUTRAL       MINION   
1060                     NAX8_03      NEUTRAL       MINION   
1061                   LOE_089t3      NEUTRAL       MINION   
1062                    TU4e_001       HUNTER         HERO   
1063                     GVG_031        DRUID        SPELL   
1064                     EX1_558      NEUTRAL       MINION   
1065                     EX1_116      NEUTRAL       MINION   
1066  TB_MechWar_Boss1_HeroPower      NEUTRAL   HERO_POWER   
1067                   LOEA07_09      NEUTRAL       MINION   
1068                      AT_114      NEUTRAL       MINION   
1069           TB_PickYourFate_6      NEUTRAL        SPELL   
1070                    AT_074e2      PALADIN  ENCHANTMENT   
1071            KARA_09_07heroic      NEUTRAL        SPELL   
1072                      AT_021      WARLOCK       MINION   
1073                    GAME_004      NEUTRAL  ENCHANTMENT   
1074              AT_132_SHAMANb       SHAMAN       MINION   
1075                     BRM_014       HUNTER       MINION   
1076                    CFM_617e      NEUTRAL  ENCHANTMENT   
1077                   LOEA16_3e      NEUTRAL  ENCHANTMENT   
1078                     CS2_121      NEUTRAL       MINION   

                         name         set  \
0               Astral Portal        KARA   
1           Ancient Teachings     EXPERT1   
2          Druid of the Flame         BRM   


Note that the results of this sort of join are dependent upon the position of each table--if you were to make `cards_with_mechanics_df` the right table and `play_requirements_df` the left table and then perform a **_Right Join_**, our results would be the same. 

**_Question:_**

Describe what was included from each table in this join.

Write your answer below this line:
________________________________________________________________________________

Every record from `cards_with_mechanics_df`, as well as any records from `play_requirements_df` that have matching index values with a record from the left table.  

#### Outer Joins

In the cell below, perform an outer join between `cards_df` and `dust_df`. Since these tables contain columns with the same name, we'll need to specify a suffix for at least one of them, so that the column can be renamed to avoid a naming collision. 

During this join, set the `rsuffix` parameter to `_dust`

In [ ]:
outer_join_df = None
outer_join_df

Inspect the output above.  Note that the naming collision has been avoided by renaming the `cost` column from the right table to `cost_dust`.  

## Summary

In this lab, you learned how to:

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)